conjure-go-contract/codecs/snappy.go

```diff
@@ -0,0 +1,88 @@
+// Copyright (c) 2022 Palantir Technologies. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package codecs
+
+import (
+	"bytes"
+	"io"
+
+	"github.com/golang/snappy"
+	werror "github.com/palantir/witchcraft-go-error"
+)
+
+var _ Codec = codecSNAPPY{}
+
+// Snappy wraps an existing Codec and uses snappy with no-framing for
+// compression and decompression using github.com/golang/snappy.
+// Ref: https://github.com/google/snappy/blob/main/format_description.txt
+func Snappy(codec Codec) Codec {
+	return &codecSNAPPY{contentCodec: codec}
+}
+
+type codecSNAPPY struct {
+	contentCodec Codec
+}
+
+func (c codecSNAPPY) Accept() string {
+	return c.contentCodec.Accept()
+}
+
+func (c codecSNAPPY) Decode(r io.Reader, v interface{}) error {
+	data, err := io.ReadAll(r)
+	if err != nil {
+		return err
+	}
+
+	decoded, err := snappy.Decode(nil, data)
+	if err != nil {
+		return err
+	}
+	return c.contentCodec.Decode(bytes.NewReader(decoded), v)
+}
+
+func (c codecSNAPPY) Unmarshal(data []byte, v interface{}) error {
+	decoded, err := snappy.Decode(nil, data)
+	if err != nil {
+		return err
+	}
+	return c.contentCodec.Unmarshal(decoded, v)
+}
+
+func (c codecSNAPPY) ContentType() string {
+	return c.contentCodec.ContentType()
+}
+
+func (c codecSNAPPY) Encode(w io.Writer, v interface{}) error {
```

Comment:
for streaming, we can use a snappy.Writer but it has to be flushed and closed:
```go
func (c codecSNAPPY) Encode(w io.Writer, v interface{}) (rErr error) {
	sw := snappy.NewBufferedWriter(w)
	defer func() {
		if err := sw.Close(); err != nil && rErr == nil {
			rErr = werror.Wrap(err, "failed to close snappy writer")
		}
	}()
	return c.contentCodec.Encode(sw, v)
}
```

Reply:
That snappy.Writer is using [framed](https://github.com/google/snappy/blob/main/framing_format.txt) compression which is different from unframed snappy compression. I actually had an implementation that used this Writer but Prometheus couldn't uncompress the framed message because it doesn't support this optional part of the standard. 

https://github.com/palantir/conjure-go-runtime/pull/326/commits/0c4540a486dc3f1fbf965864000238041405e5bd